### ISSUE TITLE
Error building from VS 2017 community 15.8.2

### ISSUE THREAD

**Issue by dazinator** [stazz/UtilPack#30]
Hello!

I am not sure if this is a utilpack issue or a VS issue.

Here is a repro csproj file, you may need to add the appveyor nuget feed: https://ci.appveyor.com/nuget/gitversion-8nigugxjftrw to restore.

```
<Project Sdk="Microsoft.NET.Sdk">

   <PropertyGroup>
      <TargetFramework>net46</TargetFramework>
  </PropertyGroup> 
  
  <ItemGroup>
       
    <PackageReference Include="GitVersionTask" Version="4.0.0-beta0015">
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup> 


</Project>

```

And a global.json:

```
{
  "sdk": {
    "version": "2.0.0"
  }
}
```

`dotnet build` seems to succeed, and I see the restore happening as well, and the task runs ok.

However building from VS 2017 Community (15.8.2) fails:

> Target "GetVersion" in file "C:\Users\darre\.nuget\packages\gitversiontask\4.0.0-beta0015\build\functionality\GitVersionBuild.targets":
1>        Initializing task factory "UtilPack.NuGet.MSBuild.NuGetTaskRunnerFactory" from assembly "C:\Users\darre\.nuget\packages\utilpack.nuget.msbuild\2.6.0\build\\net45\UtilPack.NuGet.MSBuild.dll".
1>        Detected current NuGet framework to be ".NETFramework,Version=v4.6.1", with RID "win10-x86", and local repositories: C:\Users\darre\.nuget\packages\.
1>        [NuGet Minimal]: Restoring packages for C:\Users\darre\AppData\Local\Temp\rrucvajc.g0q\dummy...
1>        [NuGet Verbose]: Restoring packages for .NETFramework,Version=v4.6.1...
1>        [NuGet Verbose]: Resolving conflicts for .NETFramework,Version=v4.6.1...
1>        [NuGet Verbose]: Checking compatibility of packages on .NETFramework,Version=v4.6.1.
1>        [NuGet Debug]: Checking compatibility for Restoring: (GitVersionTask, [4.0.0-beta0015, )) 1.0.0 with .NETFramework,Version=v4.6.1.
1>        [NuGet Debug]: Checking compatibility for GitVersionTask 4.0.0-beta0015 with .NETFramework,Version=v4.6.1.
1>        [NuGet Debug]: Checking compatibility for UtilPack.NuGet.MSBuild 2.6.0 with .NETFramework,Version=v4.6.1.
1>        [NuGet Verbose]: All packages and projects are compatible with .NETFramework,Version=v4.6.1.
1>        Using "WriteVersionInfoToBuildLog" task from the task factory "NuGetTaskRunnerFactory".
1>        Task "WriteVersionInfoToBuildLog"
1>          Task Parameter:SolutionDirectory=C:\Users\darre\Documents\Visual Studio 2017\Projects\ClassLibrary7
1>          Task Parameter:NoFetch=false
1>          Resolved GitVersionTask, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null located in C:\Users\darre\.nuget\packages\gitversiontask\4.0.0-beta0015\build\net461\GitVersionTask.dll and loaded from same path.
1>          Resolved Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a located in C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Build.dll and loaded from same path.
1>          Resolved Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a located in C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Build.dll and loaded from same path.
1>          MSBUILD : error MSB4166: Child node "2" exited prematurely. Shutting down. Diagnostic information may be found in files in "C:\Users\darre\AppData\Local\Temp\" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory.
1>          Using "GetVersion" task from the task factory "NuGetTaskRunnerFactory".
1>          Task "GetVersion"
1>            Task Parameter:SolutionDirectory=C:\Users\darre\Documents\Visual Studio 2017\Projects\ClassLibrary7
1>            Task Parameter:NoFetch=false
1>          Done executing task "GetVersion" -- FAILED.
1>        Done building target "GetVersion" in project "ClassLibrary7.csproj" -- FAILED.

As per the above, I found the MSBuild_*.failure.txt file, and this is its contents:

> UNHANDLED EXCEPTIONS FROM PROCESS 20896:
=====================
29/08/2018 00:21:44
System.ArgumentNullException: Value cannot be null.
Parameter name: value
   at System.IO.BinaryWriter.Write(String value)
   at Microsoft.Build.Framework.LazyFormattedBuildEventArgs.WriteToStream(BinaryWriter writer)
   at Microsoft.Build.Framework.BuildMessageEventArgs.WriteToStream(BinaryWriter writer)
   at Microsoft.Build.Shared.LogMessagePacketBase.WriteToStream(INodePacketTranslator translator)
   at Microsoft.Build.Shared.LogMessagePacketBase.Translate(INodePacketTranslator translator)
   at Microsoft.Build.BackEnd.NodeEndpointOutOfProcBase.RunReadLoop(Stream localReadPipe, Stream localWritePipe, ConcurrentQueue`1 localPacketQueue, AutoResetEvent localPacketAvailable, AutoResetEvent localTerminatePacketPump)
===================

UtilPack Variables:

>UtilPackNuGetMSBuildAssemblyPath = C:\Users\darre\.nuget\packages\utilpack.nuget.msbuild\2.6.0\build\\net45\UtilPack.NuGet.MSBuild.dll
1>UtilPackNuGetMSBuildPropsAbsolutePath = C:\Users\darre\.nuget\packages\utilpack.nuget.msbuild\2.6.0\build\UtilPack.NuGet.MSBuild.props
1>UtilPackNuGetMSBuildPropsPath  = C:\Users\darre\.nuget\packages\gitversiontask\4.0.0-beta0015\build\../../../utilpack.nuget.msbuild/2.6.0/build/UtilPack.NuGet.MSBuild.props
1>UtilPackNuGetMSBuildToolsRootPath = C:\Users\darre\.nuget\packages\utilpack.nuget.msbuild\2.6.0\build\\
1>UtilPackNuGetMSBuildToolsSubPath = net45\
1>UtilPackTaskFactoryParametersXML = 
1>    
```
  <NuGetFramework xmlns="http://schemas.microsoft.com/developer/msbuild/2003">.NETFramework</NuGetFramework><NuGetFrameworkVersion xmlns="http://schemas.microsoft.com/developer/msbuild/2003">4.6.1</NuGetFrameworkVersion>
1>      <PackageIDIsSelf xmlns="http://schemas.microsoft.com/developer/msbuild/2003">true</PackageIDIsSelf><AssemblyPath xmlns="http://schemas.microsoft.com/developer/msbuild/2003">GitVersionTask.dll</AssemblyPath>
```
>1>UtilPackVersion                = 2.6.0

**Comment by stazz** [comment]
Interesting, something seems to be `null` in `BuildMessageEventArgs` which shouldn't. However, is that a real cause or just a symptom of something, I do not know right now. The NuGet and VS team seem to be doing some breaking changes in almost every update, even minor, so I wouldn't be surprised if something has changed in the internals of VS and this is the result.

I need to look into this on weekend - thanks for providing repro, I will try to reproduce this with that. :)

**Comment by dazinator** [comment]
Thanks. If you need to install a previous version of VS2017 here is the link in case its handy: https://docs.microsoft.com/en-us/visualstudio/productinfo/installing-an-earlier-release-of-vs2017

**Comment by stazz** [comment]
Possibly because of the commit that affected this issue (I noticed I accidentally was giving `null` as arguments array to `BuildMessageEventArgs`), the errors I was getting in the `.txt` files with this repro were about `BuildWarningEventArgs` instead of `BuildMessageEventArgs`. However, I only use `BuildWarningEventArgs` in exactly one place, and even after replacing all `null` parameters to the constructor with empty string, I still kept getting this error.

Looking at the disassembled code in `C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Build.Framework.dll`, I noticed that `Microsoft.Build.Framework.LazyFormattedBuildEventArgs` (the source of the error) has unchecked write call to `BinaryWriter` in its `WriteToStream` method, like this: `writer.Write(this.originalCultureName);`. That field is assigned in constructor like this: `this.originalCultureName = CultureInfo.CurrentCulture.Name;`.

I am *suspecting* that current culture name is maybe `null` in some of processes started by VS...? I could try setting the `originalCultureName` field to some string via reflection a bit later to see if this is really the culprit for this crash.

EDIT: Actually, I made it throw an exception before constructing `BuildWarningEventArgs`, and the exception was not showing. So this might be some other warning coming from other component which causes the crash.
